In [1]:
from IPython.core.display import HTML
def css_styling():
    styles = open("./styles/custom.css", "r").read()
    return HTML(styles)
css_styling()

__PULL__ the changes you made at home to your local copy on the M drive.

__PULL__ the homework solutions to your local copy on the M drive.

If you need a reminder of how to do this:

Open Jupyter notebook:
<br> Start >> Programs >> Programming >> Anaconda3 >> JupyterNotebook
<br>(Start >> すべてのプログラム >> Programming >> Anaconda3 >> JupyterNotebook)

Navigate to where your interactive textbook is stored.

Open __S1_Introduction_to_Version_Control__.  

Open Jupyter notebook:
<br>Start >> Programs >> Programming >> Anaconda3 >> JupyterNotebook
<br>(Start >> すべてのプログラム >> Programming >> Anaconda3 >> JupyterNotebook)

Navigate to where your interactive textbook is stored.

Open __Seminar 5__ by clicking on __5_Library_Functions__.


# Library Functions
# Lesson Goal

 - To source and incorporate appropriate functions from external libraries to optimise your code.  





# Objectives

- Introduce use of standard library functions
- Importing and using modules
- Understanding module documentation
- Using imported functions to optimise the code you have written so far.
- Determining optimal solutions by timing your code.

Why we are studying this:
    
- To optimise your code (faster, more concise) by incorporating library functions.
- To learn to independently source functions using the internet. 
- To learn how and when to replace user defined code with appropriate library functions. 

__Lesson structure:__ Today we will learn about functions that we import from external files or “libraries”. 
 - Learn new skills together:
  - Searching for appropriate functions to do specific tasks.
  - Incorporating library functions into your code to replace larger sections of user-defined code.  
  - This first part of the practise will be done as a group (guided examples on slides and interactively completing your textbooks, feedback using your whiteboards). 
 - Review excercises (__Homework:__ Complete review exercises).  
 - Update GitHub repository. 
 - __Summary__ and __quiz__. 

## Libraries

Python, like other modern programming languages, has an extensive *library* of built-in functions. 

These functions are designed, tested and optimised by the developers of the Python langauge.  

We can use these functions to make our code shorter, faster and more reliable.


You are already familiar with some *built in* Python functions:

   - `print()` takes the __input__ in the parentheses and __outputs__ a visible representation.
   - `len()` takes a data structure as __input__ in the parentheses and __outputs__ the number of items in the data structure (in one direction).
   - `sorted()` takes a data structure as __input__ in the parentheses and __outputs__ the data structure sorted by a rule determined by the data type.
   - `abs()` takes a numeric variable as  __input__ in the parentheses and __outputs__ the mathematical absolute value of the input.

These functions belong to Python's __standard library__...

## The Standard Library

Python has a large standard library. 

It is simply a collection of Python files called 'modules'.

These files are installed on the computer you are using.

Each module contains code very much like the code that you have been writing, defining various functions. 

There are multiple modules to keep the code sorted and well organised. 

The standard libary contains many useful functions. 

They are listed on the Python website:
https://docs.python.org/3/library/functions.html

If you want to do something, for example a mathematical operation, it worth trying an internet search for a built-in function already exists.



For example, a quick google search for "python function to sum all the numbers in a list"...

<br>
https://www.google.co.jp/search?q=python+function+to+sum+all+the+numbers+in+a+list&rlz=1C5CHFA_enJP751JP751&oq=python+function+to+sum+&aqs=chrome.0.0j69i57j0l4.7962j0j7&sourceid=chrome&ie=UTF-8

...returns the function `sum()`.



`sum()` finds the sum of the values in a data strcuture. 

In [2]:
print(sum([1,2,3,4,5]))

print(sum((1,2,3,4,5)))

a = [1,2,3,4,5]
print(sum(a))



15
15
15


The function `max()` finds the maximum value in data structure.

In [3]:
print(max([4,61,12,9,2]))

print(max((3,6,9,12,15)))

a = [1,2,3,4,5]
print(max(a))

61
15
5


Two widely used packages for mathematics, science and engineeirng are `NumPy` and `SciPy`.

These are already installed on your computers.

### Importing a Package

To use an installed package, we  simply `import` it. 

## Packages

The standard library tools are available in any Python environment.

More specialised libraries are available. 

We call these packages. 



Packages contain functions and constants for more specific tasks e.g. solving trigonometric functions. 

We simply install the modules on the computer where we want to use them. 

In [4]:
import numpy 

x = 1

y = numpy.cos(x)

print(y)

print(numpy.pi)

0.540302305868
3.141592653589793


The `import` statement must appear before the use of the package in the code.  

        import numpy 

After this, any function in `numpy` can be called as:

        `numpy.function()`
        
and, any constant in `numpy` can be called as:

        `numpy.constant`.

There are a many mathematical functions available. <br>
https://docs.scipy.org/doc/numpy-1.13.0/reference/routines.math.html

## Reading function documentation

To check how to use a function e.g.:
 - what arguments to include in the () parentheses
 - allowable data types to use as arguments
 - the order in which arguments should be given 
 
search for the documentation online.

https://docs.scipy.org/doc/numpy-1.13.0/reference/routines.math.html



For example, the documentation for the function numpy.cos https://docs.scipy.org/doc/numpy-1.13.0/reference/routines.math.html includes:
    
>numpy.cos(x, /, out=None, *, where=True, casting='same_kind', order='K', dtype=None, subok=True[, signature, extobj]) 
 
> Cosine element-wise.

>x : array_like
><br>Input array in radians.

This tells us:
 - What the function does. 
 - how to call the function: 
  - we must set one function argument, `x`
  - `x` is the input to the cosine function, in radians
  - there are several default arguments (where, casting etc) that we can optionally set. These are listed after the `/` slash.
 - x should be "arraylike" (it can be an `int`, `float`, `list` or `tuple`)

Let's look at the function numpy.degrees:
https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.degrees.html

What does the function do?

What __arguments__ does it take (and are there any default arguments)? 

How would we __write__ the function when __calling__ it (accept defaults)?

What __data type__ should our input be? 

We can change the name of a package e.g. to keep our code short and neat.

Using the __`as`__ keyword:

In [5]:
import numpy as np

x = 1

y = np.cos(x)

print(y)

0.540302305868


## Namespaces
<br>By prefixing `cos` with `np`, we are using a *namespace* (which in this case is `np`).



The namespace shows we want to use the `cos` function from the Numpy package.

If `cos` appears in more than one package we import, then there will be more than one `cos` function available.

We must make it clear which `cos` we want to use. 




Often, functions with the same name, from different packages, will use a different algorithms for performing the same or similar operation. 

They may vary in speed and accuracy. 

In some applications we might need an accurate method for computing the square root, for example, and the speed of the program may not be important. For other applications we might need speed with an allowable compromise on accuracy.


Below are two functions, both named `sqrt`. 

Both functions compute the square root of the input.

 - `math.sqrt`, from the package, `math`, gives an error if the input is a negative number. It does not support complex numbers.
 - `cmath.sqrt`, from the package, `cmath`, supports complex numbers.


In [6]:
import math
import cmath

print(math.sqrt(4))
#print(math.sqrt-5)
#print(cmath.sqrt(-5))

# if we use a function name with more than one definition we get a clash
#print(sqrt(-5))

2.0


Two developers collaborating on the same program might choose the same name for two functions that perform similar tasks. 

If these functions are in different modules, there will be no name clash since the module name provides a 'namespace'. 

## Importing a Function
Single functions can be imported without importing the entire package e.g. use:

        from numpy import cos

instead of:

        import numpy 

After this you call the function without the numpy prefix: 

In [7]:
from numpy import cos

cos(x)

0.54030230586813977

In [8]:
from cmath import sqrt
from math import sqrt

#sqrt(-1)

We can even rename individual functions or constants when we import them:

In [9]:
from numpy import cos as cosine

cosine(x)

0.54030230586813977

In [10]:
from numpy import pi as pi
pi

3.141592653589793

This can be useful when importing functions from different modules:

In [11]:
from math import sqrt as square_root
from cmath import sqrt as complex_square_root

print(square_root(4))
print(complex_square_root(-1))

2.0
1j


Function names should be chosen wisely.
 - relevant
 - concise

<a id='UsingPackageFunctions'></a>
## Using Package Functions. 

Let's learn to use `numpy` functions in our programs. 





To check how to use a function e.g.:
 - what arguments to include in the () parentheses
 - allowable data types to use as arguments
 - the order in which arguments should be given 
 
look at the Numpy documentation.

A google search for 'numpy functions' returns:

https://docs.scipy.org/doc/numpy-1.13.0/reference/routines.math.html

(this list is not exhaustive). 

In [12]:
# Some Numpy functions with their definitions as given in the documentation

x = 1
y = 2
z = 3

# Trigonometric sine, element-wise.
print(np.sin(x))

# Compute tangent element-wise.
print(np.tan(x))

# Trigonometric inverse tangent
print(np.arctan(x))

# Convert angles from radians to degrees
degrees = np.degrees(x)
print(degrees)

# Convert angles from degrees to radians
radians = np.radians(degrees)
print(radians)   

0.841470984808
1.55740772465
0.785398163397
57.2957795131
1.0


__Try it yourself:__
<br> Find a function in the Python Numpy documentation that matches the function definition and use it to solve the following problem:   

Given the “legs” of a right triangle, return its hypotenuse.<br> If  the lengths of the two shorter sides of a right angle triangle are 6 units  and 3 units, what is the length of the hypotenuse?

In [13]:
# The “legs” of a right triangle are are 6 units and 3 units, 
# Return its hypotenuse in units.

Numpy functions often appear within user defined functions e.g.:

$f(x)= \cos(x) \qquad x <0$

$f(x) = \exp(-x) \qquad x \ge 0$

In [14]:
def f(x):
    if x < 0:
        f = np.cos(x)
    else:
        f = np.exp(-x)
    return f

print(f(np.pi))
print(f(np.pi/6))

0.0432139182638
0.592384847188


Package functions can be passed to other functions as arguments.

Recall __Seminar 4, What can be passed as a function argument?__



<a id='is_positive'></a>
Example: the function `is_positive` checks if the value of a function $f$, evaluated at $x$, is positive.
<br> The arguments are:
 - the function $f$
 - the value of $x$,in $f(x)$

In [15]:
def is_positive(f, x):

    if f(x) > 0:
        return True
    else:
        return False
    
def f0(x):
    """
    Computes x^2 - 1
    """
    return x*x - 1
    
# Value of x to test
x = 0.5

# Test sign of function f0 (user defined)
print(is_positive(f0, x))

# Test sign of function np.cos (numpy function)
print(is_positive(np.cos, x))

False
True


__Try it yourself:__
<br> Search online for the numpy function for each of the following mathematical functions: 
- $f = arcsin(x)$
- $f = \sqrt x$

<br> In the cell below: use the function `is_positive` to test the sign of output of the functions.  

In [16]:
# Test sign of numpy function for arcsin(x)

# Test sign of numpy function for square root of x


##### Try it yourself
In the cell below, copy and paste the `bisection` function you wrote for __Seminar 4: Review Excercise: Using Functions as Function Arguments.__

Demonstrate that your `bisection` function works correctly by finding the zero of the Numpy cos($x$) function that lies in the interval $x_1=0$ to $x_2=3$. 

In [17]:
# Bisection

## Using Package Functions to Optimise your Code

We will take previous excercises that you have completed  and look at how we can optimise them using Numpy functions.

<br> If you *have not* completed the excercises mentiond in previous seminars  you can:
1. complete the exercise from the previous seminar (without Numpy functions) then optimise using Numpy.
1. complete the excercise using Numpy functions (without completing the preceding excercise).  

Refer to your answer to __Seminar 4: Return Arguments__. 

We are going to re-write the function `compute_max_min_mean` using Numpy:


In [18]:
def compute_max_min_mean(x0, x1, x2):
    "Return maximum, minimum and mean values"
    
    x_min = x0
    if x1 < x_min:
        x_min = x1
    if x2 < x_min:
        x_min = x2

    x_max = x0
    if x1 > x_max:
        x_max = x1
    if x2 > x_max:
        x_max = x2

    x_mean = (x0 + x1 + x2)/3    
        
    return x_min, x_max, x_mean


xmin, xmax, xmean = compute_max_min_mean(0.5, 0.1, -20)
print(xmin, xmax, xmean)

-20 0.5 -6.466666666666666


Could be re-written as:

In [19]:
def np_compute_max_min_mean(x0, x1, x2):
    "Return maximum, minimum and mean values"
    
    x_min = np.amin([x0, x1, x2])
    x_max = np.amax([x0, x1, x2])
    x_mean = np.mean([x0, x1, x2])
            
    return x_min, x_max, x_mean


xmin, xmax, xmean = np_compute_max_min_mean(0.5, 0.1, -20)
print(xmin, xmax, xmean)

-20.0 0.5 -6.46666666667


### Data Structures as Function Arguments. 
Notice that the Numpy functions `amin`, `amax` and `amean` take lists as argumets.

We could simplify further by giving a single list as the argument to the function. 

This way, we can give any number of values and the function will return the aximum, minimum and mean values.

(There are alternative ways of doing this that we will study later in the course).

In [20]:
import numpy as np
def np_compute_max_min_mean(x_list):
    "Return maximum, minimum and mean values"
    
    x_min = np.amin(x_list)
    x_max = np.amax(x_list)
    x_mean = np.mean(x_list)
            
    return x_min, x_max, x_mean


xmin, xmax, xmean = np_compute_max_min_mean([0.5, 0.1, -20])
print(xmin, xmax, xmean)


print(np_compute_max_min_mean([-2, -1, 3, 5, 12]))


xmin, xmax, xmean = np_compute_max_min_mean([3, 4])
print(xmin, xmax, xmean)

-20.0 0.5 -6.46666666667
(-2, 12, 3.3999999999999999)
3 4 3.5


<a id='ElementwiseFunctions'></a>
### Elementwise Functions
Numpy functions often operate *elementwise*. 
<br> This means if the argument is a list, they will perform the same function on each element of the list.

For example, to find the square root of each number in a list, we can use:

In [21]:
a = [9, 25, 36]
print(np.sqrt(a))

[ 3.  5.  6.]


<a id='MagicFunctions'></a>
### Magic Functions
We can use *magic function* (http://ipython.readthedocs.io/en/stable/interactive/magics.html), `%timeit`, to compare the time the user-defiend function takes to execute compared to the Numpy function. 

(e.g. Sometimes we must choose between minimising the length of the code and minimising the time it takes to run.) 

In [22]:
%timeit compute_max_min_mean(0.5, 0.1, -20)
print("")
%timeit np_compute_max_min_mean(0.5, 0.1, -20)

The slowest run took 7.65 times longer than the fastest. This could mean that an intermediate result is being cached.
1000000 loops, best of 3: 411 ns per loop



TypeError: np_compute_max_min_mean() takes 1 positional argument but 3 were given

##### Try it yourself 
In the cell below, find a Numpy function that provides the same solution as the function your write as your answer to __Seminar 3, Review Exercise: Indexing, part (A)__: 
<br>Add two vectors, $\mathbf{A}$ and $\mathbf{B}$ such that:
$ \mathbf{A} + \mathbf{B} = [(A_1 + B_1), 
                              (A_2 + B_2),
                              ...
                              (A_n + B_n)]$

__(A)__ Use the Numpy function to add vectors:

$\mathbf{A} = [-2, 1, 3]$

$\mathbf{B} = [6, 2, 2]$

Check that your answer is the same as your answer to __Seminar 3, Review Exercise: Indexing__.

__(B)__ Using your answer to  __Seminar 3, Review Exercise: Indexing__ write a function `vector_add` that takes vectors A and B as inputs and returns the sum of the two vectors by calling:

```python
vector_add(A, B)
```

__(C)__ Use *magic function* `%timeit`, to compare the spped of the Numpy function to the user defined function `vector_add`. 
<br> Which is fastest?

In [ ]:
# Vector Addition

<a id='ImportingAlgorithms'></a>
## Importing Algorithms as Functions (e.g. Root finding)

So far we have mostly looked at library functions that perform single mathematical operations such as trigonomtric or algebraic functions. 

Library functions also include those that can be used for complete multi-stage tasks.



For example, in place of the `bisection` function you wrote to find the root of a function, a number of root-finding functions from imported modules can be used. 

The package `scipy.optimize` contains a number of functions for estimating the roots of a function including:
 - `scipy.optimize.bisect`
 - `scipy.optimize.fsolve` (the most popular root-finding function)

The documentation for `fsolve` https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.optimize.fsolve.html:

>scipy.optimize.fsolve(func, x0, args=(), fprime=None, full_output=0, col_deriv=0, xtol=1.49012e-08, maxfev=0, band=None, epsfcn=None, factor=100, diag=None)[source]

>Return the roots of the (non-linear) equations defined by func(x) = 0 given a starting estimate.

>__func__ : callable f(x, *args)
<br> &nbsp; &nbsp; &nbsp; A function that takes at least one (possibly vector) argument.
<br>__x0__ : ndarray
<br>&nbsp; &nbsp; &nbsp; The starting estimate for the roots of func(x) = 0.

What does the function do?

What __arguments__ does it take (and are there any default arguments)? 

How would we __write__ the function when __calling__ it (accept defaults)?

What __data type__ should our input be? 

__Bisection method:__ the user selects the interval in which to look for the root. 

__`solve` method:__ the user selects an *initial estimate* for the root.



To demonstrate, here is an example:
    
The function, $f(x) = x^3 + 4x^2 + x - 6$ has roots -3, -2, and 1. 

The function should return the root that is closest to our estimate. 

In [ ]:
import scipy
from scipy.optimize import fsolve

def func(x):
    return x**3 + 4*x**2 + x - 6

a = scipy.optimize.fsolve(func, -30)

print(a)

__Try it yourself:__
<br> In the cell below, use `scipy.optimize.fsolve()` to print the root of cos(x) (using `np.cos()`).
<br> Try several initial guess values of x.  

In [ ]:
# Find the root of cos(x)

Sometimes, we want to find more than one root of a function. 

 - For the function $cos(x)$, finding all roots is impractical as our solution would be infinite.
 <br>

 - For functions like the polynomial $f(x) = x^3 + 4x^2 + x - 6$ we can use the function `np.roots()` to find all roots.
 <br> The function argument is the coeffients of the polynomial as a list.

In [ ]:
print(np.roots([1, 4, 1, -6]))

__Try it yourself:__
<br>In the cell below use `numpy.roots()` to find the roots of the function:
<br>$y = x^3 - 2x^2 - 11x + 12$

In [ ]:
# root finding using numpy.roots()

## Review Exercises
The following excercises will help you to practise finding useful functions form external packages and applying them when solving engineering problems. 

### Review Exercise: Numpy Package Functions. 
<br> Find a function in the Python Numpy documentation that matches the function definition and use it to solve the problems below:

__(A)__ Calculate the exponential of all elements in the input array.
<br> Print a list where each element is the exponential of the corresponding element in list a:
<br>`a = [0.1, 0, 10]`

In [ ]:
# Print a list where each element is the exponential of the corresponding element in list a

__(B)__ Convert angles from degrees to radians..
<br> Convert angle `theta`, expressed in degrees, to radians:
<br>`theta` = 47

In [ ]:
# convert angle `theta`, expressed in degrees, to radians

### Review Exercise: Searching for Appropriate Package Functions 
<br>
Refer to your answer to __Seminar 4, Review Exercise: Default Arguments.__
Copy and paste your code in the cell below.

__(A)__ *Elementwise functions* perform an operation on each element of a data structure.
<br>Within the function create a list to store the values x, y and z:
```python
def magnitude(x, y, z = 0):
    """
    Returns the magnitude of a 2D or 3D vector
    """
    vector = [x, y, z]
```

 Within your function, replace the operation for square (raise to power of 2) $^2$  with an elementwise numpy function that takes the list `vector` as an argument. 
 
 <a href='#ElementwiseFunctions'>Jump to Elementwise Functions</a>

__(B)__ Find an Numpy functions to the replace operation for:

 - summation $\sum$
 - square root $\sqrt x$
and include these in your function. 

__(C)__ Use *magic function* `%timeit`, to compare the speed of your user-defined function (from Seminar 4) to the speed when using Numpy functions.
<br> Which is fastest?

 <a href='#MagicFunctions'>Jump to Magic Functions</a>

__(D)__ Search online for a single numpy function that takes a vector as input and returns the magnitide of a vector. 
<br> Use it calculate the magnitude of the vector $x$. 
<br> Check the answer against the value generated in __A__
<br> Check your answers using hand calculations.

__(E)__ Use *magic function* `%timeit`, to compare the time for:
 - the Numpy function to return the magnitude
 - the function you used in parts __(A)-(C)__ 
for 2D and 3D vectors. 


In [ ]:
# Searching for Appropriate Package Functions 

### Review Exercise: Using Package Functions to Optimise your Code

Search for a Numpy function that has a __similar__ function to the `is_positive` function from Section: <a href='#UsingPackageFunctions'>Using Package Functions</a>; the answer it returns should show if an input value is positive or not. 

In the cell below:
 - copy and paste the `is_positive` function
 - use the magic function %timeit to compare the speed of the `is_positive` function with the Numpy function for analysing the sign of a numerical input.

<a href='#is_positive'>Jump to function:`is_positive`</a> 



In [ ]:
np.sign(-2)

### Review Exercise: Alternative Expressions
Recall __Seminar 3, Indexing__. 

We saw that the __dot product__ of two vectors can be experssed both geometrically and algebraically. 

__GEOMETRIC REPRESENTATION__

\begin{align}
\mathbf{A} \cdot \mathbf{B} = |\mathbf{A}| |\mathbf{B}| cos(\theta)
\end{align}

__ALGEBRAIC REPRESENTATION__

>So the dot product of two 3D vectors:
> <br> $ \mathbf{A} = [A_x, A_y, A_z]$
> <br> $ \mathbf{B} = [B_x, B_y, B_z]$
> <br> is:

\begin{align}
\mathbf{A} \cdot \mathbf{B} &= \sum_{i=1}^n A_i B_i \\
&= A_x B_x + A_y B_y + A_z B_z.
\end{align}



In the cell titled " `The dot product of C and D `", you wrote a program to compute the dot product using:
 - a for loop
 - indexing 

$\mathbf{C} = [2, 4, 3.5]$

$\mathbf{D} = [1, 2, -6]$


In the cell below, use:
 - the Numpy cosine function
 - the magnitude function that you used in the last example (either  user defined or Numpy function)
 
to compute $\mathbf{C} \cdot \mathbf{D}$ using the geomtric expression.

Check your answer is the same as your answer from Seminar 3. 



### Review Exercise: Importing Algorithms as Functions

In <a href='#ImportingAlgorithms'>Importing Algorithms as Functions (e.g. Root finding)</a> we learnt that the package scipy.optimize contains a number of functions for estimating the roots of a function, including `scipy.optimize.bisect`.

This function performs the same/ a similar function to the `bisection` function that you have been developing. 

__(A)__ Find the documentation for the function `scipy.optimize.bisect` to learn how to use it.

__(B)__ Use `scipy.optimize.bisect` to estimate the root of the function $f(x) = 2sin^2 x - 3sin x + 1$:
<br> &nbsp; &nbsp; &nbsp; (i) between 0 and $\frac{\pi}{6}$
<br> &nbsp; &nbsp; &nbsp; (ii) between 1.5 and 2
<br> &nbsp; &nbsp; &nbsp; (iii) between $\frac{3}{4}\pi$ and $\pi$

__NOTE:__ &nbsp; $sin^2(x) = (sin(x))^2$

__(C)__ Use the magic function %timeit to compare the speed of your user-sefined function `bisection`, with the speed of `scipy.optimize.bisect`. 

# Updating your git repository

You have made several changes to your interactive textbook.

The final thing we are going to do is add these changes to your online repository so that:
 - I can check your progress
 - You can access the changes from outside of the university server. 
 
 > Save your work.
 > <br> `git add -A`
 > <br>`git commit -m "A short message describing changes"`
 > <br>`git push origin master`
 
 <br>Refer to supplementary material: __S1_Introduction_to_Version_Control.ipynb__.  

# Summary

- Python has an extensive __standard library__ of built-in functions. 
- More specialised libraries of functions and constants are available. We call these __packages__. 
- Packages are imported using the keyword ....
- The function documentation tells is what it does and how to use it.
- When calling a library function it must be prefixed with a __namespace__ is used to show from which package it should be called.  
- The magic function .... can be used to time the execution of a function. 




# Homework 

1. __PULL__ the changes you made in-class today to your personal computer.
1. __COMPLETE__ any unfinished Review Exercises.
1. __PUSH__ the changes you make at home to your online repository. 

<br>Refer to supplementary material: __S1_Introduction_to_Version_Control.ipynb__.  